google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java

```diff
@@ -432,13 +444,26 @@ private void publishAllWithoutInflightForKey(final String orderingKey) {
   }
 
   private ApiFuture<PublishResponse> publishCall(OutstandingBatch outstandingBatch) {
-    return publisherStub
-        .publishCallable()
-        .futureCall(
-            PublishRequest.newBuilder()
-                .setTopic(topicName)
-                .addAllMessages(outstandingBatch.getMessages())
-                .build());
+    if (enableCompression && outstandingBatch.batchSizeBytes >= compressionBytesThreshold) {
```

Comment:
We should be able to simplify this a bit by creating the default context above, adding the compress call option if required, and then making a unified call to the publisherStub with either the default or updated context.

Reply:
@mmicatka I had same suggestion in my design doc, but then I was discouraged to make any change to the existing flow, that is, not to pass default context to the current futureCall.  And I think that suggestion makes sense as it is a critical path.

Reply:
Makes sense to me.

Reply:
We could still create the context once and reuse it across the calls to publish though, couldn't we?

Reply:
done!

Reply:
I don't think what I expected to happen here is exactly what happened. I was thinking that context could be a member variable of the class itself. And you could also have compressedContext be a member variable that sets compression (if enableCompression is true). Then, you can choose which context to use in the publish method based on whether or not you should compress. I don't know if there is any overhead in calling GrpcCallContext.createDefault(), but by doing it once at the class level, you'll avoid any overhead that does exist.

Reply:
Unresolving due to @kamalaboulhosn followup

Reply:
Thank Mike! I have created a class variable publishContext to remove the overhead of GrpcCallContext.createDefault() with every publish call. However, we have to add gzip compression to the context with every publish call as it depends on the message size and the compression threshold.

Reply:
Why can't you create a second context called `publishContextWithCompression` and then choose between the two of them based on whether or not you should compress? As you have it now, won't publishing any message that gets compressed result in all subsequent messages being compressed because you change the call option in `publishContext` only when you enable compression?